### PR TITLE
AssemblyResolve handler for plugin third-party dependencies

### PR DESCRIPTION
**Pull request by hal0x2328**
I'm writing a [plugin for neo-cli to publish smart contract events to a Redis PubSub queue](https://github.com/hal0x2328/NeoPubSub). I quickly ran into a problem - since neo-cli plugins themselves are loaded at runtime, the existing .NET mechanism for resolving third-party dependencies in assemblies does not work (at least not for .Net Core), and the plugin fails to load with the following error: 
```
System.IO.FileNotFoundException: Could not load file or assembly 'ServiceStack.Redis, Version=5.0.0.0, Culture=neutral,
 PublicKeyToken=null'. The system cannot find the file specified.
```
This patch attempts to address this by adding an [AssemblyResolve](https://docs.microsoft.com/en-us/dotnet/api/system.appdomain.assemblyresolve?view=netframework-4.7.2) handler that is called when an assembly fails to load, and proceeds to attempt to resolve and load the dependencies of the plugin. (Note that the way the `CurrentDomain_AssemblyResolve` function is written here, the third-party DLLs the plugin is dependent on are expected to be in the main working directory of neo-cli, it may be the case that we want to place them in a subfolder of the plugin's tree).

After this patch is applied I no longer experience the above error and my plugin functions as expected.

In my research on this topic, I've also discovered that the approach I've taken may not be the ideal way to go about solving the problem - some other issues with runtime Assembly loading are noted [on this page](https://natemcmaster.com/blog/2018/07/25/netcore-plugins/) such as versioning issues/race conditions that might occur if multiple plugins attempt to use different versions of the same third-party library. This may not be a huge concern at present, so instead of incorporating this entire third-party class I implemented the more straightforward single-function solution.

### CODE REVIEW

File: neo/Plugins/Plugin.cs [hunk 1]
```diff
@@ -95,7 +97,10 @@ internal static void LoadPlugins(NeoSystem system)
                     {
                         constructor?.Invoke(null);
                     }
-                    catch { }
+                    catch (Exception ex)
+                    {
+                        Console.WriteLine("Failed to initialize plugin: " + ex);
```
Comment (erikzhang) [inline neo/Plugins/Plugin.cs]:
```suggestion
                        Plugin.Log(nameof(Plugin), LogLevel.Error, $"Failed to initialize plugin: {ex.Message}"); 
```

File: neo/Plugins/Plugin.cs [hunk 2]
```diff
@@ -120,5 +125,26 @@ public static bool SendMessage(object message)
...
+        private static Assembly CurrentDomain_AssemblyResolve(object sender, ResolveEventArgs args)
+        {
+            if (args.Name.Contains(".resources"))
+                return null;
+
+            Assembly assembly = AppDomain.CurrentDomain.GetAssemblies().FirstOrDefault(a => a.FullName == args.Name);
+            if (assembly != null)
+                return assembly;
+
+            string filename = args.Name.Split(',')[0] + ".dll";
```
Comment (shargon) [inline neo/Plugins/Plugin.cs]: What format do you receive here in `args.Name`?

File: neo/Plugins/Plugin.cs (hunk 1 above)
Comment (hal0x2328) [reply neo/Plugins/Plugin.cs]: Implemented suggested change in both exception handlers https://github.com/neo-project/neo/pull/521/commits/b6e4a999fe232d4f2b3cf86e7a814ff789bdc7e0

File: neo/Plugins/Plugin.cs (hunk 2 above)
Comment (hal0x2328) [reply neo/Plugins/Plugin.cs]: It's receiving the full assembly name, e.g.
```MyAssembly, Version=1.2.3.4, Culture=neutral, PublicKeyToken=null```

Comment (erikzhang) [reply neo/Plugins/Plugin.cs]: You can use `AssemblyName` to parse the full name of the assembly:

```c#
AssemblyName a = new AssemblyName(args.Name);
string fullname = a.FullName;
string name = a.Name;
```

Comment (hal0x2328) [reply neo/Plugins/Plugin.cs]: Neat, just made that change in https://github.com/hal0x2328/neo/commit/6160d4cf119e92911d7df1166309a5b25a4eb556

**Pull request merged**